image/cas/interface.go

```diff
@@ -0,0 +1,39 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package cas implements generic content-addressable storage.
+package cas
+
+import (
+	"io"
+)
+
+// Engine represents a content-addressable storage engine.
+type Engine interface {
+
+	// Put adds a new blob to the store.
+	Put(writer io.Reader) (digest string, err error)
```

Comment:
On Thu, Jul 07, 2016 at 02:53:36PM -0700, Stephen Day wrote:
> 1. It's much easier to control the copy if you have something to
>    write to, rather than passing in a writer and hoping for the best.
> 2. Consider the case of re-entrant writes. I don't think they are
>    necessary here, but having a repeatable identifier for resuming a
>    failed write exists there.

Yeah, I think that's more than we want to bite off here.  And I'm not
sure how much of that is useful to expose to the generic caller.  For
example, you could handle resumable uploads *inside* the engine.
You'd need to hash the content internally to know which ID to pick
from your cache of previously-started-and-failed pushes, but I'm not
sure how you avoid that by exposing the create/resume information to
the caller.  I'll dig into the docker/distribution code and see where
it's getting the resume IDs from to confirm this approach.

> 3. Consider the transaction boundary at the point this write
>    completes. This is interesting when two writers are trying to
>    write the same blob. This must ensure that only one succeeds but
>    they may race each other.

Why does only one have to succeed?  If the blob being put is now in
the store, that sounds like success to me (so idempotent Put).  It's
probably worth documenting that so callers are all on the same page;
I'll push an update shortly.

> 4. Consider the use of handles, rather than directly accessing the
>    readers and writers. This will allow you to have better control
>    over the scope of a reader and writer.
> 5. Be very careful with `ReadFrom`, if you rely on it for
>    performance…

These sound like details for production engine implementations, and I
think the image-spec implementations should try and be simple and
robust reference implementations (even if that means they are slow and
inefficient).  Folks who want a production engine should be writing
that in a separate repository, and they can put their wrapper around
the Reader/Writer inside the engine without changing the Engine
interface.

> 6. Consider splitting the front end interface from the driver. Make
>    the front end concrete. There are many details that are important
>    for the driver that aren't important for the user of the
>    interface. (We made this mistake in distribution).

I'm not sure what you mean by “make the frontend concrete”, but the
goal of this PR is to hide all those messy details behind a generic
interface.

> Camlistore also is a great resource for structuring these kinds of
> interfaces, as well.

The resumable Camlistore interface seems to be here [1], and uses an
opaque continuation token.  I'm not clear on how you recover tokens
after failed pushes, but will look more later.

Both the docker/distribution and Camilstore interfaces take Context
arguments [2] for easy canceling.  I'll mimic that here (and replace
the ‘done’ from refs.Engine.List).

[1]: https://godoc.org/camlistore.org/pkg/blobserver#BlobStreamer
[2]: https://godoc.org/golang.org/x/net/context#Context